### PR TITLE
Added spawnpoints-only mode to beehive

### PR DESCRIPTION
**Pull request by Xcelled**
## Description
Beehive can be quite inefficient, scanning over and over areas that have no spawnpoints. I've implemented this PR to automatically remove areas without spawnpoints from beehive's searches.

## Motivation and Context
Implemented as a proof of concept of filtering a search location based on spawnpoints. This is a useful proof of concept, though, in that it significantly reduces the work needed to scan an area with beehive, once spawnpoints have been collected. It's kind of a hybrid beehive/spawnpoint scanner that's a quick way to get some benefits of spawnpoint scanning.

When using this option, it's possible to reduce the number of accounts or increase the scan delay time and still scan the same area in the same time. Useful if you're being rate-limited or live in a rural area.

Also, generated locations are now cached in a list instead of being re-computed every time.

## How Has This Been Tested?
I ran both in spawnpoints mode and regular mode.

## Screenshots (if appropriate):
image

image

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by stlmikey80** [comment]
How do I use this?

**Comment by chuyskywalker** [comment]
https://github.com/PokemonGoMap/PokemonGo-Map/issues/179#issuecomment-239650626

**Comment by danieliser** [comment]
Awesome.

**Comment by stlmikey80** [comment]
Very nice, i like it..

**Comment by chuyskywalker** [comment]
Works for me, doesn't impede any existing functionality, heading down the path we're going towards. May be removed, when better more integrated feature set comes along, but functional for now.

2016-08-13 22_03_08-pokemon go map

An example of SF where I only let the regular beehive run for one loop, then activated this mode. Did was what expected.

**Comment by Xcelled** [comment]
Cute graphic :3

File: pogom/search.py [hunk 1]
```diff
@@ -161,11 +163,32 @@ def search_overseer_thread(args, new_location_queue, pause_bit, encryption_lib_p
...
+                log.debug('Spawnpoint search radius: %f', sp_dist)
+                # generate coords of the midpoints of each edge of the square
+                south, west = get_new_coords(current_location, sp_dist, 180), get_new_coords(current_location, sp_dist, 270)
+                north, east = get_new_coords(current_location, sp_dist, 0), get_new_coords(current_location, sp_dist, 90)
+                # Use the midpoints to arrive at the corners
+                log.debug('Searching for spawnpoints between %f, %f and %f, %f', south[0], west[1], north[0], east[1])
+                spawnpoints = set((d['latitude'], d['longitude']) for d in Pokemon.get_spawnpoints(south[0], west[1], north[0], east[1]))
+                if len(spawnpoints) == 0:
+                    log.warning('No spawnpoints found in the specified area! (Did you forget to run a normal scan in this area first?)')
+                locations = [coords for coords in locations if any(geopy_distance.distance(coords, x).meters <= 70 for x in spawnpoints)]
```
Comment (chuyskywalker) [inline pogom/search.py]: This is...a lot of list comprehension and I'm not python-ista enough yet to grok this at a glance. It's working, but confusing to me.

Comment (Xcelled) [inline pogom/search.py]: Blame flake! The lambda version was easier to understand.

```python
any_spawnpoints_in_range = lambda coords: any(geopy_distance.distance(coords, x).meters <= 70 for x in spawnpoints)
locations = [x for x in locations if any_spawnpoints_in_range(x)]
```
>any will return True when at least one of the elements is Truthy.
It's also short circuiting, so it stops at the first True.

The argument we pass to any builds a generator (**not** a list) for on demand evaluation. This generator operates on the spawnpoints set and, for each one, returns if it is less than 70 meters away from the given coords. The net result being that that lambda returns true if at least one spawn point is in range of the circle location.

That lambda is used inside a list comprehension that essentially acts as a filter. It returns a list of all locations where that location has a spawnpoint in range.

Flake complained about the lambda, so I had to replace the friendly name with the actual lambda, but it works the same.

Comment (chuyskywalker) [inline pogom/search.py]: Huh, well here's why it complained: http://stackoverflow.com/questions/25010167/e731-do-not-assign-a-lambda-expression-use-a-def

_Kinda_ makes sense....I guess. ^shrug^ As with JS, we picked a common standard and just went with it in order to just keep things inline.

Comment (Xcelled) [inline pogom/search.py]: Yeah... I agree for larger/reused functions, you definitely want to def them (seewhatIdidthere). But in this case, my purpose was to decompose one line into two for readability, and adding a def would have worsened that by making you track down another function def that would have been just one line used nowhere else... Actually... wow I'm ashamed of myself. I screwed this up in terms of readability. One sec, I'll fix it in a new PR.

Comment (Xcelled) [inline pogom/search.py]: Done, see #645. I forgot I could put a def there :<

**Comment by FrostTheFox** [comment]
only one dev +1? niiiice

**Comment by uparikh113** [comment]
Might be missing something really obvious but ran the scan with 0 steps and spawn points only. Got the error that no spawn points were found even though I ran the beehive scan. Do I need to extract out the spawns from the database myself and specify them in a specific file?

**Comment by drewreece** [comment]
@uparikh113, you are misunderstanding what this does - do not use a step size of 1. This code only removes cells without ANY spawnpoints, it doesn't use the timing data at all. Go and use another fork if you want to grab spawns points that are currently spawning. e.g. Sorokyls/ TBTerra versions in #179 or maybe #585.

**Comment by ovkhan** [comment]
How do i enable this mode?
I don't see anything in options ...

**Comment by vinsilsim** [comment]
another legal option would enable a flag to search in all areas with points of spawn.

**Comment by drewreece** [comment]
@ovkhan re-read the options (python runserver.py --help).  If you do not have '--spawnpoints-only' you need to git fetch & pull or check what version you are running.

**Comment by drewreece** [comment]
@vinsilsim, what do you mean? The new '--spawnpoints-only' option sort of does that already & also removes empty points from scans.

**Comment by Becold** [comment]
Does it scan pokestop & gyms too ? It doesnt seems to

**Comment by vinsilsim** [comment]
@drewreece this option makes the search go in all areas with marked respawn? without putting coordinates?

**Comment by drewreece** [comment]
Maybe I'm being dumb - I still don't understand vinsilsim, this is the 'standard beehive pattern' that is defined in your config or command line arguments. It simply excludes cells that have never shown spawns in earlier scans.

Are you talking about a different change? Try the feature to see what it does, pass --spawnpoints-only to the command.

**Comment by drewreece** [comment]
@Becold, I see Pokéstops & gyms using this feature. Lured stops change status, gyms seem to be correct.

You will need to provide more info or raise an issue if you cannot see them, anyway github is not for support so I'm going to stop trying to answer support questions here.

**Comment by lungbubble** [comment]
Just copied those 2 files, and when finally doing --scanpoints-only I got 
AttributeError: type object 'Pokemon' has no attribute 'get_spawnpoints'

**Comment by Xcelled** [comment]
You can't just drop these files in. This has been merged into develop - just pull the latest down.

**Comment by rrbailey89** [comment]
@Xcelled will this scan less often but more effectively .. or will this cycle the scan rotation again faster resulting in scanning the same locations more often?

**Comment by Xcelled** [comment]
The latter. I'd recommend increasing your sd so you don't get banned for teleporting.

**Comment by rrbailey89** [comment]
@Xcelled do you recommend using an -SD larger than 10?

**Comment by Xcelled** [comment]
Yes. How much larger, though, is up to you, and depends on how many spawnpoints you have nearby.

**Comment by rrbailey89** [comment]
@Xcelled [this is the area im covering](http://goo.gl/bFScVP) I suppose ill try 15 or 20 maybe

**Comment by Xcelled** [comment]
Sounds like a good staring point yeah. You can tweak it later if need be.

**Pull request merged**